tensorpack/tfutils/symbolic_functions.py

```diff
@@ -212,3 +212,149 @@ def saliency_map(output, input, name="saliency_map"):
     saliency_op = tf.gradients(max_outp, input)[:][0]
     saliency_op = tf.identity(saliency_op, name=name)
     return saliency_op
+
+
+def contrastive_loss(left, right, y, margin, extra=False):
+    """Loss for Siamese networks as described in the paper:
+    `Learning a Similarity Metric Discriminatively, with Application to Face
+    Verification <http://yann.lecun.com/exdb/publis/pdf/chopra-05.pdf>` by Chopra et al.
+
+    .. math::
+        \min 0.5 * Y * D^2 + 0.5 * (1-Y) * \{\max(0, m - D)\}^2, D = \Vert l - r \Vert_2
+
+    Args:
+        left (tf.Tensor): left of image pair
+        right (tf.Tensor): right of image pair
+        y (tf.Tensor): label (1: similar, 0:not similar)
+        margin (float): horizont for negative examples (y==0)
+        extra (bool, optional): return distances for pos and negs
+
+    Returns:
+        tf.Tensor: constrastive_loss, [pos_dist, neg_dist]
+    """
+    with tf.name_scope("constrastive_loss"):
+        y = tf.cast(y, tf.float32)
+
+        diff = left - right
+
+        delta = tf.reduce_sum(tf.square(diff), 1)
+        delta_sqrt = tf.sqrt(delta + 1e-6)
+
+        match_loss = delta
+        missmatch_loss = tf.square(tf.nn.relu(margin - delta_sqrt))
+
+        loss = tf.reduce_mean(0.5 * (y * match_loss + (1 - y) * missmatch_loss))
+        pos_dist = tf.reduce_mean(y * delta_sqrt, name="pos-dist")
+        neg_dist = tf.reduce_mean((1 - y) * delta_sqrt, name="neg-dist")
+
+        if extra:
+            return loss, pos_dist, neg_dist
+        else:
+            return loss
+
+
+def cosine_loss(left, right, y):
+    """Loss for Siamese networks (cosine version).
+
+    Remarks:
+        Same as `contrastive_loss` but with different similarity measurment.
+
+    Args:
+        left (tf.Tensor): left of image pair
+        right (tf.Tensor): right of image pair
+        y (tf.Tensor): label (1: similar, 0:not similar)
+
+    Returns:
+        tf.Tensor: cosine-loss as scalar (and) average_pos_loss, average_neg_loss
+    """
+
+    def l2_norm(t, eps=1e-12):
+        """return L2 norm for input x
+
+        Args:
+            t (tf.Tensor): input tensor
+            eps (float, optional): constant for numerical stability
+
+        Returns:
+            tf.Tensor: norm of input tensor
+        """
+        with tf.name_scope("l2_norm"):
+            return tf.sqrt(tf.reduce_sum(tf.square(t), 1, True) + eps)
+
+    with tf.name_scope("cosine_loss"):
+        y = 2 * tf.cast(y, tf.float32) - 1
+        pred = tf.reduce_sum(left * right, 1) / (tf.squeeze(l2_norm(left) * l2_norm(right)) + 1e-10)
+
+        return tf.nn.l2_loss(y - pred)
```

Comment:
So cosine_loss is summed over the batch, but others are averaged over the batch? This is inconsistent and will confuse people when they want to interpret the magnitude of loss.